myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,747 @@
+            + np.random.normal(0, noise)
+        )
+    # Discard the warmup draws
+    return draws[warmup:]
+
+
+ar1_data = simulate_ar(10, -0.9, 0)
+
+fig, ax = plt.subplots(figsize=(10, 3))
+ax.set_title("Generated Autoregressive Timeseries", fontsize=15)
+ax.plot(ar1_data);
+```
+
+## Specifying the Model
+
+We'll walk through the model step by step and then generalise the pattern into a function that can be used to take increasingly complex structural combinations of components.
+
+```{code-cell} ipython3
+## Set up a dictionary for the specification of our priors
+## We set up the dictionary to specify size of the AR coefficients in
+## case we want to vary the AR lags.
+priors = {
+    "coefs": {"mu": [10, 0.2], "sigma": [0.1, 0.1], "size": 2},
+    "sigma": 8,
+    "init": {"mu": 9, "sigma": 0.1, "size": 1},
+}
+
+## Initialise the model
+with pm.Model() as AR:
+    pass
+
+## Define the time interval for fitting the data
+t_data = list(range(len(ar1_data)))
+## Add the time interval as a mutable coordinate to the model to allow for future predictions
+AR.add_coord("obs_id", t_data, mutable=True)
+
+with AR:
+    ## Data containers to enable prediction
+    t = pm.MutableData("t", t_data, dims="obs_id")
+    y = pm.MutableData("y", ar1_data, dims="obs_id")
+
+    # The first coefficient will be the constant term but we need to set priors for each coefficient in the AR process
+    coefs = pm.Normal("coefs", priors["coefs"]["mu"], priors["coefs"]["sigma"])
+    sigma = pm.HalfNormal("sigma", priors["sigma"])
+    # We need one init variable for each lag, hence size is variable too
+    init = pm.Normal.dist(
+        priors["init"]["mu"], priors["init"]["sigma"], size=priors["init"]["size"]
+    )
+    # Steps of the AR model minus the lags required
+    ar1 = pm.AR(
+        "ar",
+        coefs,
+        sigma=sigma,
+        init_dist=init,
+        constant=True,
+        steps=t.shape[0] - (priors["coefs"]["size"] - 1),
+    )
+
+    # The Likelihood
+    outcome = pm.Normal("likelihood", mu=ar1, sigma=sigma, observed=y)
+    ## Sampling
+    idata_ar = pm.sample_prior_predictive()
+    idata_ar.extend(pm.sample(2000, random_seed=100, target_accept=0.95))
+    idata_ar.extend(pm.sample_posterior_predictive(idata_ar))
+```
+
+```{code-cell} ipython3
+idata_ar
+```
+
+Lets check the model structure with plate notation and then examine the convergence diagnostics.
+
+```{code-cell} ipython3
+az.plot_trace(idata_ar, figsize=(10, 6), kind="rank_vlines");
+```
+
+Next we'll check the summary estimates for the to AR coefficients and the sigma term.
+
+```{code-cell} ipython3
+az.summary(idata_ar, var_names=["~ar"])
+```
+
+```{code-cell} ipython3
+fig, ax = plt.subplots(figsize=(10, 4))
+idata_ar.posterior.ar.mean(["chain", "draw"]).plot(ax=ax, label="Posterior Mean AR level")
+ax.plot(ar1_data, "o", color="black", markersize=2, label="Observed Data")
+ax.legend()
+ax.set_title("Fitted AR process \n and observed data");
```

Comment:
Very pedantic point, but I don't think we need the spaces around the \n. It acts to make the title off centre.

Reply:
No problem, will adjust.